### DISCUSSION TITLE
Keycloak performance

### DISCUSSION THREAD

**Discussion by hnhai** [keycloak/keycloak#14221] [Archive]
- I'm install Keycloak 18.0.2 by Docker (Server 8 CPU, 16Gb ram) connect to PostgresDB in another Sever
- LoadTest By Jmeter call to rest api of Keycloak to get Jwt token (1000 CCUs, duration: 3 minutes)
- Keycloak used almost 8 core, response very slow. Start from miliseconds but increment to 30 seconds.
It is normal ? How I can fix it ? Thanks

**Comment by hmlnarik** [comment]
Please provide more detailed description of your scenarios and deployment setup.

**Reply by hnhai** [reply]
- Setup: Linux server (8 core, 16gb ram) with docker, using [https://github.com/bitnami/containers/blob/main/bitnami/keycloak/docker-compose.yml](url) image. Postgres DB in anothor Server
- Load Test: Jmeter using csv file of username and password, running 1000 Threads, duration 3 minutes. Call api login to Keycloak IP 
Actual: Low throughput about 80 tps. High reponse time, avg: 20 seconds
Expect: High throughput, low response time < 2 seconds

Hi @hmlnarik, is this enough ? Thanks

**Reply by hmlnarik** [reply]
Which "API login" endpoint does the test call?
Do all the "1000 threads" hit the same single server at the same time?

**Reply by hnhai** [reply]
@hmlnarik 
> Which "API login" endpoint does the test call?
This API: /realms/realm_name/protocol/openid-connect/token

> Do all the "1000 threads" hit the same single server at the same time?
Yes

**Reply by hnhai** [reply]
Hi @hmlnarik do you have any recomend ?

**Reply by hmlnarik** [reply]
- "Username and password" seems to point to resource owner grant flow. This is relies on hashing the passwords which is intentionally CPU-intensive, this requirement being independent of IAM implementation, Keycloak or other. However in real-world scenario, it is more likely that the users reuse their established sessions rather than reauthenticating with every request.
- 1000 simultaneous threads means lots of context switching (OS-dependent, Keycloak independent), and that applies to the load driver as well.
- There was no mention of Keycloak being warmed up before the benchmark actually starts. Is there any warm-up time for JVM?

In total, I am afraid that the setup does not reflect realistic scenarios. Consider checking [keycloak-benchmark](https://github.com/keycloak/keycloak-benchmark/) project for assessing the resources needed to achieve the throughput for your requirements.

**Reply by hnhai** [reply]
@hmlnarik I got it. Thank for your help

**Reply by nikhilmalhotra55** [reply]
@hnhai how did you fix it ? facing the same problem. Around 350-400ms response time for raw api call on keycloak. same specs as yours.
I built a wrapper auth microservice which uses keycloak internally, but login api takes very long when using username, password based, token based api's are still fast.
